File: EDUCATIONAL105/c.py
#!/usr/bin/env python
import os
import sys
from io import BytesIO, IOBase
import math
from collections import Counter
import bisect
 
 
def func1(array_a, array_b):
    if (not array_a) or not array_b:
        return 0
    max_count = 0
    curr = 0
    b_start = 0
    b_end = 0
    # print(array_a, array_b)
    good_pos = set(array_a).intersection(set(array_b))
    good_len = len(good_pos)
    while b_end < len(array_b):
        # print("here", curr, b_start, b_end)
        while curr < len(array_a) and array_a[curr] <= array_b[b_end]:
            if array_a[curr] in good_pos:
                good_len -= 1
            curr += 1
        # print("curr", curr, good_len)
        while b_start < b_end and array_b[b_start] + curr <= array_b[b_end]:
            # print(array_b[b_start], array_b[b_end])
            b_start += 1
        max_count = max(max_count, min(b_end - b_start + 1, curr) + good_len)
        # print("b start end", b_start, b_end)
        b_end += 1
        # print(max_count)
    print(max_count)
    return max_count
 
 
def func(array_a, array_b):
    a_0 = bisect.bisect_left(array_a, 0)
    # print('a_0 = ', a_0, 'array_a = ', array_a)
    b_0 = bisect.bisect_left(array_b, 0)
    return func1(array_a[a_0:], array_b[b_0:]) + func1(
        [-i for i in array_a[:a_0][::-1]], [-i for i in array_b[:b_0][::-1]]
    )
 
 
def main():
    num_test = int(parse_input())
    for _ in range(num_test):
        n, m = [int(i) for i in parse_input().split()]
        array_a = [int(i) for i in parse_input().split()]
        array_b = [int(i) for i in parse_input().split()]
        print(func(array_a, array_b))
 
parse_input = lambda: sys.stdin.readline().rstrip("\r\n")
 
# endregion
 
if __name__ == "__main__":
    main()

File: EDUCATIONAL105/c.js
'use-strict'
 
process.stdin.resume();
process.stdin.setEncoding('utf8');
 
let data = '';
let currentLine = 0;
 
process.stdin.on('data', chunk => {
data += chunk;
});
 
process.stdin.on('end', () =>{
    data = data.replace(/\s*$/, '')
    .split('\n')
    .map(str => str.replace(/\s*$/, ''));
 
    let testCases = parseInt(data.shift());
 
    while(testCases--) {
        const [n,m] = readLine().split(' ').map(Number);
        const a = readLine().split(' ').map(Number);
        const b = readLine().split(' ').map(Number);
        const a0 = a.filter(num => num < 0);
        const a1 = a.filter(num => num >= 0);
        const b0 = b.filter(num => num < 0);
        const b1= b.filter(num => num >= 0);
        const a0_filter = a0.filter(num => num >= b0[0]);
        const b0_filter = b0.filter(num => num <= a0_filter[a0_filter.length - 1])
        const a00_filter = a0_filter.map(num => num *-1);
        const b00_filter = b0_filter.map(num => num *-1);
        const a1_filter = a1.filter(num => num <= b1[b1.length - 1]);
        const b1_filter = b1.filter(num => num >= a1_filter[0]);
        console.log(count1(a00_filter.sort((a,b) => a - b), b00_filter.sort((a,b) => a - b),1) + count1(a1_filter, b1_filter,1) );
    }
});
 
function get_ints() { 
    return data.shift().split(' ').map(Number);
}
 
function readLine() { 
        return data[currentLine++];
}
 
function intersection(setA, setB) {
    let _intersection = new Set()
    for (let elem of setB) {
        if (setA.has(elem)) {
            _intersection.add(elem)
        }
    }
    return _intersection
}

function count1(a,b,opt = 0){
    if(a.length === 0 || b.length === 0) return 0;
    let globalMax = 0;
    const setB = new Set([...b]);
    const setA = new Set([...a]);
    let good_pos = intersection(setA, setB)
    let good_len = good_pos.size;
    const start_time = new Date();
    let j =0;
    let i = 0;
    let b_start = 0;
    while(i < b.length){
        // console.log('here', 'curr = ', j, 'b_start  = ', b_start, 'b_end = ', i)
        while(j < a.length && a[j] <= b[i]){
            // console.log(`curr = ${j}, good_len = ${good_len}`)
            if(good_pos.has(a[j])){
                good_len -= 1;
            }
            j += 1;
        }
        while(b_start < i && b[b_start] + j <= b[i]){
            b_start += 1
        }

        globalMax = Math.max(globalMax, Math.min(i - b_start + 1, j) + good_len);
        // console.log(`globalMax+ = ${globalMax}`)
        i++;
    }
    // console.log(`finish in: ${new Date() - start_time} with ${globalMax} `)
    return globalMax;
}
 
module.exports = count1